projects/limtoc/main.cc

```diff
@@ -0,0 +1,223 @@
+/*
+    ChibiOS - Copyright (C) 2006..2015 Giovanni Di Sirio
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include "ch.h"
+#include "hal.h"
+
+#include "analog.h"
+#include "encoder.h"
+
+#include "gitsha1.h"
+#include "blink.h"
+#include "usbconfig.h"
+#include "saconfig.h"
+#include "utility.h"
+
+#include "parameters.h"
+
+#include "discrete_linear.h"
+#include "kalman.h"
+
+namespace {
+    constexpr systime_t loop_period = MS2ST(1);
+    constexpr float dt = static_cast<float>(ST2MS(loop_period))/1000.0f; // seconds
+    constexpr float k = 3.14f; // N-m/rad
+    constexpr float m = 1.0f; // kg-m^2
+
+    // sensors
+    Analog analog;
+    Encoder encoder_steer(sa::RLS_ROLIN_ENC, sa::RLS_ROLIN_ENC_INDEX_CFG);
+
+    dacsample_t set_handlebar_reference(float reference) {
+#if defined(LIMTOC_VELOCITY_MODE)
+        constexpr float MAX_REF_VALUE = sa::MAX_KOLLMORGEN_VELOCITY;
+#else // defined(LIMTOC_VELOCITY_MODE)
+        constexpr float MAX_REF_VALUE = sa::MAX_KOLLMORGEN_TORQUE;
+#endif // defined(LIMTOC_VELOCITY_MODE)
+        const float saturated_reference = util::clamp(reference, -MAX_REF_VALUE, MAX_REF_VALUE);
+        const dacsample_t aout = saturated_reference/MAX_REF_VALUE*sa::DAC_HALF_RANGE + sa::DAC_HALF_RANGE;
```

Comment:
`constexpr` is a good idea.

`adcsample_t` is a typedef-ed as `uint16_t` but doesn't use the whole range as the ADCs are 12-bit.